File: 0731-my-calendar-ii/0731-my-calendar-ii.cpp
class MyCalendarTwo {
public:
   vector<pair<int,int>> booking,overlapping;
    MyCalendarTwo() {
        
    }
    
    bool book(int s, int e) {
        for (auto& [start,end]: overlapping){
            if (s< end && e> start)return false;
        }
        for (auto& [start,end] : booking){
            if (s< end && e > start)overlapping.push_back({max(s, start), min(e, end)});
        }
        booking.emplace_back(s,e);
        return true;
    }
};

/**
 * Your MyCalendarTwo object will be instantiated and called as such:
 * MyCalendarTwo* obj = new MyCalendarTwo();
 * bool param_1 = obj->book(start,end);
 */

File: 0731-my-calendar-ii/0731-my-calendar-ii.java
class MyCalendarTwo {
    private List<int[]> book,over;
    public MyCalendarTwo() {
        book = new ArrayList<>();
        over = new ArrayList<>();
    }
    
    public boolean book(int start, int end) {
        for (int[] o: over){
            if (start< o[1] && end> o[0])return false;
        }
        for (int[] b : book){
            if (start < b[1] && end> b[0]){

                over.add(new int[]{Math.max(start,b[0]),Math.min(end,b[1])});
            }
        }
        book.add(new int[]{start,end});
        return true;
    }
}

/**
 * Your MyCalendarTwo object will be instantiated and called as such:
 * MyCalendarTwo obj = new MyCalendarTwo();
 * boolean param_1 = obj.book(start,end);
 */